### ISSUE TITLE
JPA static metamodel generation error: cannot find symbol

### ISSUE THREAD

**Issue by willcrisis** [jhipster/generator-jhipster#6673]
##### **Overview of the issue**

After generated an entity and selected filtering option, I can't compile and run project anymore. Application uses Gradle and MySQL database for production, and H2 database for development.

Output of console is:

```
PS D:\des\rota> .\gradlew
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileJava
Note: Hibernate JPA 2 Static-Metamodel Generator 5.2.10.Final
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:68: error: cannot find symbol
                specification = specification.and(buildSpecification(criteria.getId(), ContaPagar_.id));
                                                                                       ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:71: error: cannot find symbol
                specification = specification.and(buildRangeSpecification(criteria.getVencimento(), ContaPagar_.vencimento));
                                                                                                    ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:74: error: cannot find symbol
                specification = specification.and(buildRangeSpecification(criteria.getValor(), ContaPagar_.valor));
                                                                                               ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:77: error: cannot find symbol
                specification = specification.and(buildSpecification(criteria.getStatus(), ContaPagar_.status));
                                                                                           ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:80: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getTipoItemId(), ContaPagar_.tipoItem, TipoItem_.id));
                                                                                                              ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:80: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getTipoItemId(), ContaPagar_.tipoItem, TipoItem_.id));
                                                                                                                                    ^
  symbol:   variable TipoItem_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:83: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getFormaPagamentoId(), ContaPagar_.formaPagamento, FormaPagamento_.id));
                                                                                                                    ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:83: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getFormaPagamentoId(), ContaPagar_.formaPagamento, FormaPagamento_.id));
                                                                                                                                                ^
  symbol:   variable FormaPagamento_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:86: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getFornecedorId(), ContaPagar_.fornecedor, Fornecedor_.id));
                                                                                                                ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:86: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getFornecedorId(), ContaPagar_.fornecedor, Fornecedor_.id));
                                                                                                                                        ^
  symbol:   variable Fornecedor_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:89: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getVendaId(), ContaPagar_.venda, Venda_.id));
                                                                                                           ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:89: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getVendaId(), ContaPagar_.venda, Venda_.id));
                                                                                                                              ^
  symbol:   variable Venda_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:92: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getContaReceberId(), ContaPagar_.contaReceber, ContaReceber_.id));
                                                                                                                  ^
  symbol:   variable ContaPagar_
  location: class ContaPagarQueryService
D:\des\rota\src\main\java\com\willcrisis\rota\service\ContaPagarQueryService.java:92: error: cannot find symbol
                specification = specification.and(buildReferringEntitySpecification(criteria.getContaReceberId(), ContaPagar_.contaReceber, ContaReceber_.id));
                                                                                                                                            ^
  symbol:   variable ContaReceber_
  location: class ContaPagarQueryService
14 errors


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 17s
5 actionable tasks: 4 executed, 1 up-to-date

```

##### **Motivation for or Use Case**

Project is no longer usable after generated an entity and selected filtering option.

##### **Reproduce the error**

1. Generate a new JHipster project using Gradle and MySQL database for production.
2. Generate a few entities with some fields and relationships, and select filtering option.
3. Try to restart the application or run "gradlew" command.

##### **JHipster Version(s)**

4.10.2

##### **JHipster configuration**

Using JHipster version installed locally in current project's node_modules
Executing jhipster:info
Options: 
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
rota@0.0.0 D:\des\rota
`-- generator-jhipster@4.10.2

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.willcrisis.rota",
      "nativeLanguage": "pt-br"
    },
    "jhipsterVersion": "4.10.2",
    "baseName": "rota",
    "packageName": "com.willcrisis.rota",
    "packageFolder": "com/willcrisis/rota",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "no",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "gradle",
    "enableSocialSignIn": false,
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "angularX",
    "useSass": true,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "pt-br",
    "languages": [
      "pt-br",
      "en"
    ]
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity FormaPagamento (forma_pagamento) {
  nome String required minlength(1)
}
entity TipoItem (tipo_item) {
  nome String required minlength(1)
}
entity Cliente (cliente) {
  nome String required minlength(1),
  email String required minlength(1),
  telefone String,
  endereco String,
  cpfCnpj String minlength(14) maxlength(18)
}
entity Fornecedor (fornecedor) {
  nome String required minlength(1),
  cpfCnpj String,
  email String,
  telefone String,
  percentualComissao BigDecimal
}
entity Venda (venda) {
  data LocalDate required
}
entity ContaReceber (conta_receber) {
  parcelas Integer required min(1),
  vencimento LocalDate required,
  status Status required,
  valor BigDecimal required
}
entity ContaPagar (conta_pagar) {
  vencimento LocalDate required,
  valor BigDecimal required,
  status Status required
}
entity Comissao (comissao) {
  valor BigDecimal required,
  vencimento LocalDate required,
  status Status required
}

enum Status {
  ABERTO,
  QUITADO
}

relationship OneToMany {
  Venda{contasReceber} to ContaReceber{venda required},
  Venda{contasPagar} to ContaPagar{venda required}
}
relationship ManyToOne {
  Venda{cliente(nome) required} to Cliente,
  ContaReceber{formaPagamento(nome) required} to FormaPagamento,
  ContaPagar{tipoItem(nome) required} to TipoItem,
  ContaPagar{formaPagamento(nome) required} to FormaPagamento,
  ContaPagar{fornecedor(nome) required} to Fornecedor,
  ContaPagar{contaReceber} to ContaReceber,
  Comissao{contaPagar required} to ContaPagar
}

paginate FormaPagamento, TipoItem, Cliente, Fornecedor, Venda, ContaReceber, ContaPagar, Comissao with pagination
service Venda, ContaReceber, ContaPagar, Comissao with serviceClass

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

git version 2.14.1.windows.1

node: v8.4.0

npm: 5.3.0

bower: 1.8.0

gulp:
[21:25:10] CLI version 1.4.0

yeoman: 2.0.0

yarn: 1.2.1

Docker version 17.09.0-ce, build afdb6d4

docker-compose version 1.16.1, build 6d1ac219

Congratulations, JHipster execution is complete!

##### **Browsers and Operating System**

* Windows 10

- [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by kcrowe123** [comment]
I am not sure what gradle equivalent is but I had to add this to my pom.xml for maven build.
`                <plugin>
                    <groupId>org.bsc.maven</groupId>
                    <artifactId>maven-processor-plugin</artifactId>
                    <version>2.2.4</version>
                    <executions>
                        <execution>
                            <id>process</id>
                            <goals>
                                <goal>process</goal>
                            </goals>
                            <phase>generate-sources</phase>
                            <configuration>
                                <processors>
                                    <processor>org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor</processor>
                                </processors>
                            </configuration>
                        </execution>
                    </executions>
                    <dependencies>
                        <dependency>
                            <groupId>org.hibernate</groupId>
                            <artifactId>hibernate-jpamodelgen</artifactId>
                            <version>${hibernate.version}</version>
                        </dependency>
                    </dependencies>
                </plugin>`

**Comment by Tcharl** [comment]
The procedure is listed in [the doc](http://www.jhipster.tech/entities-filtering/), unformtunately, the jboss page does not list the gradle way of doing things

**Comment by gzsombor** [comment]
I guess, you don't have mapstruct, as the seemingly innocent line, which generates the metamodel classes are [here](https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/gradle/_mapstruct.gradle).

It shouldn't be too hard to fix this.

**Comment by gzsombor** [comment]
I was wrong, this works, as that 'mapstruct.gradle' is included even when mapstruct is not used.
When I start gradlew it even shows that the required metamodel generator is started:

```
➜  ./gradlew                                                                                                                                                                                                674ms
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileJava
Note: Hibernate JPA 2 Static-Metamodel Generator 5.2.10.Final

> Task :bootRun
...
```

**Comment by willcrisis** [comment]
I can see that the files are correctly generated in build/generated/source/apt/main folder, but seems that they are not added to classpath. I tried some configurations in build.gradle to add them to srcDirs like this:
```
sourceSets.main.java.srcDirs = ['build/generated/source/apt/main', 'src/main/java]
```

No success at all.

**Comment by deepu105** [comment]
@gzsombor are you able to reproduce this with the given config?

**Comment by deepu105** [comment]
@MathieuAA on a sidenote do you know why the JDL created using export JDL doesnt include the filtering option?

**Comment by gzsombor** [comment]
I've tried to reproduce it with the given jdl and yo-rc.json files, but locally it works.
When I've tried to use the system gradle, instead of the generated one, it failed, but that problem is not a silent one, but I guess, this https://github.com/tbroyer/gradle-apt-plugin/issues/66. 

Maybe you have a different gradle version?

**Comment by jdubois** [comment]
I also couldn't reproduce - I used your conf, ran Gradle, everything works fine.
So I'm closing this, I guess you have a wrong Gradle version, or maybe an IDE configuration issue? Could you try again with `./gradlew`?

**Comment by MathieuAA** [comment]
@deepu105 It may be fixed already (from JCore v1.4.1 I believe). I'll create an issue in the JCore project to keep a log of the possible issue and reproduce.
